src/nvim/api/tabpage.c

```diff
@@ -14,6 +18,47 @@
 # include "api/tabpage.c.generated.h"
 #endif
 
+/// Opens a new tabpage with a single window
+///
+/// @param buffer   Buffer handle, or 0 for current buffer
+/// @param enter    Boolean, whether to enter the new tabpage
+/// @param opts     Optional parameters
+///  - after: Tabpage handle, open new tabpage after this tabpage.
+///           Defaults to opening after the current tabpage.
+/// @param[out] err Error details, if any
+/// @return Handle to newly created tabpage
+Tabpage nvim_open_tabpage(Buffer buffer, Boolean enter, Dict(open_tabpage_opts) *opts, Error *err)
+  FUNC_API_SINCE(12)
+{
+  buf_T *buf = find_buffer_by_handle(buffer, err);
+  if (!buf) {
+    return 0;
+  }
+
+  int after = 0;
+  if (HAS_KEY(opts, open_tabpage_opts, after)) {
+    tabpage_T *tp = find_tab_by_handle((Tabpage)opts->after, err);
+    if (!tp) {
+      return 0;
+    }
+    // Add 1 to the tabpage index because of tabpage numbering.
+    // Neglecting to do this will result in the new tabpage being opened
+    // *before* the specified tabpage.
+    after = tabpage_index(tp) + 1;
+  }
+
+  tabpage_T *tp = win_new_tabpage(after, buf->b_ffname, enter);
+
+  if (!tp) {
+    api_set_error(err, kErrorTypeException, "Failed to create tabpage");
+    return 0;
+  }
+
+  win_set_buf(tp->tp_firstwin, buf, false, err);
+
+  return tp->handle;
```

Comment:
Autocommands from `win_new_tabpage`, `win_set_buf` have to be considered here:

- `tp` may have been immediately freed (even `tp->handle` isn't valid in that case).
- `buf` may have been immediately freed.
- `tp->tp_firstwin` may be wrong if `win_new_tabpage` autocommands enter the tabpage and mess with the layout.